compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp

```diff
@@ -252,6 +257,143 @@ template void hoistStaticallyBoundAllocationsInFunc<memref::AllocOp>(
+  });
+  for (auto ordinalOp : ordinalOps) {
+    int64_t ordinal = ordinalOp.getOrdinal().getSExtValue();
+    if (ordinal >= workloadVals.size()) {
+      ordinalOp.emitOpError(
+          "ordinal number is higher than the number of workloads captured in "
+          "the workgroup count region");
+    }
+    map.map(ordinalOp.getResult(),
+            workloadVals[ordinalOp.getOrdinal().getSExtValue()]);
+  }
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(workgroupCountOp);
+  for (auto op : slice) {
+    // TODO(#13038) This is a WAR for the these ops ending up in workgroup count
+    // computation. They should not. Some pre-processing at MaterializeEncoding
+    // time might make these go away.
+    if (isa<IREE::VMVX::QueryTileSizesOp>(op)) {
+      Value constVal =
+          rewriter.create<arith::ConstantIndexOp>(op->getLoc(), 16);
+      for (auto result : op->getResults()) {
+        map.map(result, constVal);
+      }
+      continue;
+    }
+    rewriter.clone(*op, map);
+  }
+  SmallVector<OpFoldResult> results;
+  // Since the workgroup count at HAL level is in x, y, z form, process the
+  // workload in reverse.
+  for (auto ofr : llvm::reverse(workgroupCount)) {
+    if (auto val = ofr.dyn_cast<Value>()) {
+      results.push_back(getAsOpFoldResult(map.lookup(val)));
+    } else {
+      results.push_back(ofr);
+    }
+  }
+
+  // The `maxWorkgroupParallelDims` represents the maximum dimension number
+  // used for distribution. The rest of the workgroups get folded into the
+  // `maxWorkgroupParallelDims`
+  Location loc = workgroupCountOp.getLoc();
+  if (results.size() > maxWorkgroupParallelDims) {
+    MutableArrayRef<OpFoldResult> resultsRef =
+        llvm::MutableArrayRef<OpFoldResult>(results);
+    assert(maxWorkgroupParallelDims != 0 &&
+           "unexpected max parallel dimensions being 0");
+    AffineExpr s0, s1;
+    bindSymbols(rewriter.getContext(), s0, s1);
+    AffineMap foldMap = AffineMap::get(0, 2, s0 * s1);
+    for (auto [index, foldedResult] : llvm::enumerate(
+             resultsRef.take_back(results.size() - maxWorkgroupParallelDims))) {
+      resultsRef[maxWorkgroupParallelDims - 1] =
+          affine::makeComposedFoldedAffineApply(
+              rewriter, loc, foldMap,
+              {resultsRef[maxWorkgroupParallelDims - 1],
+               resultsRef[maxWorkgroupParallelDims + index]});
+    }
+    results.resize(maxWorkgroupParallelDims);
+  }
+
+  // Fill out the remaining results with 1.
+  if (results.size() < workgroupCountOp.getNumResults()) {
+    results.resize(workgroupCountOp.getNumResults(), rewriter.getIndexAttr(1));
+  }
+  rewriter.replaceOp(workgroupCountOp,
+                     getValueOrCreateConstantIndexOp(rewriter, loc, results));
+  for (auto ordinalOp : ordinalOps) {
+    rewriter.replaceOp(ordinalOp, ordinalOp.getOperand());
+  }
+
+  return success();
+}
+
+LogicalResult lowerWorkgroupCountFromBodySliceOp(
```

Comment:
I think it's better to use `rewriter.notifyMatchFailure` in this method. They are more like matching failures to me. The omitOpError is more about op verification.

Reply:
I actually want this to fail if there is a problem. So thats why I am using `emitOpError`